nni/algorithms/compression/v2/pytorch/common/data_collector.py

```diff
@@ -0,0 +1,58 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import logging
+from typing import Dict, List
+
+from torch import Tensor
+
+from nni.algorithms.compression.v2.pytorch.base.pruner_tools import DataCollector, TrainerBasedDataCollector
+
+_logger = logging.getLogger(__name__)
+
+__all__ = ['WeightDataCollector', 'WeightTrainerBasedDataCollector', 'SingleHookTrainerBasedDataCollector']
+
+
+class WeightDataCollector(DataCollector):
+    """
+    Collect all wrapper weights.
+    """
+
+    def reset(self):
+        pass
+
+    def collect(self) -> Dict[str, Tensor]:
+        data = {}
+        for _, wrapper in self.compressor.get_modules_wrapper().items():
+            data[wrapper.name] = wrapper.module.weight.data.clone().detach()
+        return data
+
+
+class WeightTrainerBasedDataCollector(TrainerBasedDataCollector):
+    """
+    Collect all wrapper weights after training or inference.
+    """
+
+    def collect(self) -> Dict[str, Tensor]:
+        for _ in range(self.training_epochs):
+            self.trainer(self.compressor.bound_model, self.optimizer, self.criterion)
+
+        data = {}
+        for _, wrapper in self.compressor.get_modules_wrapper().items():
+            data[wrapper.name] = wrapper.module.weight.data.clone().detach()
+        return data
+
+
+class SingleHookTrainerBasedDataCollector(TrainerBasedDataCollector):
```

Comment:
Single can only has one hook to collect data. So it can not satisfy the pruner which calculates metric based on both activation and gradient? Would we support that?

Reply:
Yes, we can support this by adding more logic in `collect` function. due to we don't have this scenario right now, I am a little lazy to implement it...